Add: Joke module

zephiowl opened the pull request:
Added a basic joke module using an API. I tried to get it to search for if the current channel had NSFW enabled, but I couldn't figure out how to parse the data correctly on a channel level, only a guild one. Be sure to let me know if you want any changes done

zephiowl committed:
Add: basic joke module

zephiowl committed:
Add: Core functionality to joke module

zephiowl committed:
fix: remove random mistype

zephiowl committed:
Add: Better error handling

zephiowl committed:
Add: NSFW config flag

zephiowl committed:
Fix: debug log

zleyyij commented on src/modules/joke.ts:
```diff
@@ -0,0 +1,64 @@
+
+// Config manager for nsfw detection
+const jokeConfig = botConfig.modules.joke;
+let JOKE_API_URL = '';
```
Typically `SCREAMING_SNAKE_CASE`  is used for global constants, but you have something like `BASE_JOKE_API_URL`, then have a separate variable like `blacklistFlags` that's an array of flags that get joined together, eg:
```typescript
let jokeApiUrl = 'https://v2.jokeapi.dev/joke/Any';
const blacklistFlags = ['religious', 'political', 'racist', 'sexist', 'explicit'];
if (jokeConfig.nsfw) {
  blacklistFlags.push('nsfw');
}
jokeUrl += `?blacklistFlags=${blacklistFlags.join(',')}&type=single'}`;
```
This is one way you could implement it, but please don't use `SCREAMING_SNAKE_CASE` for variables that are mutated.

zleyyij commented on src/modules/joke.ts:
```diff
@@ -0,0 +1,64 @@
+
+  async (args, interaction) => {
+    fetchJoke()
+      .then(joke => {
```
Please prefer `await` to `.then()` here.
```typescript
const joke = await fetchJoke();
```
It reduces the amount of nesting, and because `await` is locally blocking, it helps keep control flow clear.

zleyyij commented on src/modules/joke.ts:
```diff
@@ -0,0 +1,64 @@
+  async (args, interaction) => {
+    fetchJoke()
+      .then(joke => {
+        if (joke == 'Failed to find joke') {
```
A strict equality operator (`===`) would be preferred here over the loose equality operator (`==`).

Xanthus58 replied:
I tried doing this as i noticed it too, but then it seemed to fail the if statement every time, not entirely sure why

zleyyij commented on src/modules/joke.ts:
```diff
@@ -0,0 +1,64 @@
+      const data = await response.json();
+      return data.joke;
+    } else {
+      return 'Failed to find joke';
```
Nit:
Rather than returning a string containing the error, it would be preferred to throw an error and let the error be handled elsewhere (where the function is called), and add a jsdoc annotation with `@throws`.
```typescript
/**
 * fetch a joke and return it yada yada
 * @throws Will throw an error if the API call fails
 */
 
// -- snip --

throw new Error('Failed to fetch joke');
 ```

This change will not be required to complete the merge, it is optional.

Xanthus58 replied:
Yeah i'll see if i can do this for ya, it was Ajax that suggested i didn't throw an error in the first place
https://discord.com/channels/1092132405428637886/1092132414228271116/1172493888179667025

Cpt-Dingus commented:
Personal nit
```typescript
  'joke', // command name
  'Get a funny joke from the bot', // command description
  [], // dependencies
  [], // options
  ```
  
  I don't think these comments should be present outside of a testing environment

zephiowl committed:
fix!: added suggestions

Now i'm running into an issue where sometimes it fetches the joke perfectly with no issues, but other times i get a `Invalid Form Body` error.

Xanthus58 commented:
Now I'm running into an issue where sometimes it fetches the joke perfectly with no issues, but other times i get a `Invalid Form Body` error from within the discordjs module itself it seems and crashes the whole bot. And I'm not entirely sure how to resolve this issue. I would appreciate any insight you might have

![image](https://github.com/r-Techsupport/TuringBot/assets/66909997/052e8e66-a985-4ca0-b6f7-7e10e01caba1)
![image](https://github.com/r-Techsupport/TuringBot/assets/66909997/05d75adf-a135-4050-927d-42743749ebd5)

zleyyij commented:
Try using `console.log` to verify that the joke is a valid string before sending it. `embeds[0].description` is the field that was not set correctly.

zephiowl committed:
Fix: Removed undefined errors

I feel so dumb, there was a tiny typo that broke the entire thing, and i forgot to format the response as a string :sob:

zleyyij commented on src/modules/joke.ts:
```diff
@@ -0,0 +1,60 @@
+);
+
+async function fetchJoke(): Promise<string> {
+  try {
```
This try/catch is redundant.

You can also catch the error on line 38 with 
```typescript
await fetchJoke().catch((error) => {
  // run your error handling code here
});
```

Xanthus58 replied:
https://github.com/r-Techsupport/TuringBot/pull/53/commits/d049ce68a93b74d6a25d356502de747bb5797bfe

zephiowl committed:
Fix: Remove try catch

@zleyyij wanted it so i implemented it

Cpt-Dingus commented:
LGTM
